### PR TITLE
Move `AbstractBuild/index.jelly` to `Run/index.jelly`

### PR DESCRIPTION
**Pull request by janfaracik**
Would be good to have some discussion over this.

This PR/discussion proposes moving the build page from `AbstractBuild` to its parent `Run`, this is so that plugins that extend `Run` can use the page rather than having to reimplement it (e.g. like [workflow-job-plugin](https://github.com/jenkinsci/workflow-job-plugin/blob/master/src/main/resources/org/jenkinsci/plugins/workflow/job/WorkflowRun/index.jelly)). 

It would be preferable to have plugins extend pages rather than to reimplement them to make changes in `core` easier and more consistent. This will be important with upcoming works on the Jenkins user experience, such as introducing app bars, new widgets and adjusting the sidebar.

For AbstractBuild I've split out Changesets/Upstream and downstream relationships into separate files, one called `summary.jelly` for use in the summary list, the other into `main.jelly`.

Searched for classes that extend `Run` with https://github.com/search?q=org%3Ajenkinsci+%22class%22+%22extends+Run%3C%22&type=code

From that search the following plugins extend `Run`:
* https://github.com/jenkinsci/workflow-job-plugin/blob/master/src/main/resources/org/jenkinsci/plugins/workflow/job/WorkflowRun/index.jelly (opened a PR [here](https://github.com/jenkinsci/workflow-job-plugin/pull/364))
* https://github.com/jenkinsci/external-monitor-job-plugin/tree/master/src/main/resources/hudson/model/ExternalRun (replaces index.jelly with console.jelly - probably fine to leave as is)
* https://github.com/jenkinsci/template-workflows-plugin/blob/ede8b144fa8e4091acca4f75eb234a937e10c595/src/main/java/org/jenkins/plugin/templateWorkflows/TemplateswWorkflowRun.java#L3 (no replacement for index.jelly)

Happy to hear thoughts on the matter (and alternatives if I'm misunderstanding how everything works).

### Testing done

* The AbstractBuild page works as before/looks identical to before.

### Proposed changelog entries

- N/A

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [ ] The Jira issue, if it exists, is well-described.
- [ ] The changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developers, depending on the change) and are in the imperative mood (see [examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)).
  - Fill in the **Proposed upgrade guidelines** section only if there are breaking changes or changes that may require extra steps from users during upgrade.
- [ ] There is automated testing or an explanation as to why this change has no tests.
- [ ] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadocs, as appropriate.
- [ ] New deprecations are annotated with `@Deprecated(since = "TODO")` or `@Deprecated(forRemoval = true, since = "TODO")`, if applicable.
- [ ] New or substantially changed JavaScript is not defined inline and does not call `eval` to ease future introduction of Content Security Policy (CSP) directives (see [documentation](https://www.jenkins.io/doc/developer/security/csp/)).
- [ ] For dependency updates, there are links to external changelogs and, if possible, full differentials.
- [ ] For new APIs and extension points, there is a link to at least one consumer.

### Desired reviewers

@jenkinsci/sig-ux 

### Maintainer checklist

Before the changes are marked as `ready-for-merge`:

- [ ] There are at least two (2) approvals for the pull request and no outstanding requests for change.
- [ ] Conversations in the pull request are over, or it is explicit that a reviewer is not blocking the change.
- [ ] Changelog entries in the pull request title and/or **Proposed changelog entries** are accurate, human-readable, and in the imperative mood.
- [ ] Proper changelog labels are set so that the changelog can be generated automatically.
- [ ] If the change needs additional upgrade steps from users, the `upgrade-guide-needed` label is set and there is a **Proposed upgrade guidelines** section in the pull request title (see [example](https://github.com/jenkinsci/jenkins/pull/4387)).
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).

<a href="https://gitpod.io/#https://github.com/jenkinsci/jenkins/pull/8110"></a>

### CODE REVIEW

**Comment by timja** [comment]
@jglick any issues with this? it sounds like a good idea to me

**Comment by timja** [comment]
@janfaracik can you file a downstream PR to workflow-job removing it's copy after checking to see if there's any differences?

**Comment by janfaracik** [comment]
> @janfaracik can you file a downstream PR to workflow-job removing it's copy after checking to see if there's any differences?

Opened https://github.com/jenkinsci/workflow-job-plugin/pull/364

**Comment by daniel-beck** [comment]
`index.jelly` lists changesets, something [available](https://javadoc.jenkins.io/hudson/model/AbstractBuild.html#getChangeSets()) for `AbstractBuild` but not `Run`. Same with upstream/downstream relationships.

Therefore of `index.jelly` by lines makes no sense for `Run`. There might be a basic implementation listing actions, but just moving doesn't seem like a clean change.

**Review by daniel-beck — DISMISSED** [review]
Per above.

**Comment by uhafner** [comment]
> `index.jelly` lists changesets, something [available](https://javadoc.jenkins.io/hudson/model/AbstractBuild.html#getChangeSets()) for `AbstractBuild` but not `Run`. Same with upstream/downstream relationships.
> 
> Therefore of `index.jelly` by lines makes no sense for `Run`. There might be a basic implementation listing actions, but just moving doesn't seem like a clean change.

In this case it would make split the index.jelly into a "frame" jelly page and multiple local parts, that can be implemented or overwritten for specific types. Would that be feasible? @janfaracik did you try to make a diff of both files? Are there other known implementations?

**Comment by janfaracik** [comment]
> Would that be feasible? 

Yeah. End goal would be a widget system where Jenkins/plugins can plug their own widgets into the page.

> @janfaracik did you try to make a diff of both files?

Yeah I went through and I believe the summary components were the only differences.

> Are there other known implementations?

I need to have a look around.

**Review by jglick — DISMISSED** [review]
What @daniel-beck said, basically. There are other things that do not align. For example the _upstream/downstream builds_ section [would not exist](https://issues.jenkins.io/browse/JENKINS-29913) for `WorkflowRun`.

If there are particular sections of the page that seem duplicated and are likely to change often, then introduce a new helper fragment for them, and consume that in `workflow-job`.

**Review by jglick — DISMISSED** [review]

File: core/src/main/resources/hudson/model/AbstractBuild/summary.jelly
Comment (janfaracik) [reply core/src/main/resources/hudson/model/AbstractBuild/summary.jelly:32]: Let me know if I'm misunderstanding -  Add a file that uses the updated changesets method, then `<st:include ... />` it in both `AbstractBuild/main.jelly` and `WorkflowRun/main.jelly`?

Comment (janfaracik) [reply core/src/main/resources/hudson/model/AbstractBuild/summary.jelly]: I've added `<st:include page="changesets.jelly" class="jenkins.scm.RunWithSCM" />` to `summary.jelly` in `AbstractBuild` and `WorkflowRun`. The implementation of changesets is different in the two files (as you noted above) with one difference being that `WorkflowRun` lacks 

```
<!-- dependency changes -->
              <j:set var="depChanges" value="${it.getDependencyChanges(it.previousBuild)}"/>
              <j:if test="${!empty(depChanges)}">
                ${%Changes in dependency}
                <ol>
                  <j:forEach var="dep" items="${depChanges.values()}">
                    <li>
                      <a href="${rootURL}/${dep.project.url}" class="model-link">${dep.project.displayName}</a>
                      <st:nbsp/>
                      <j:choose>
                        <j:when test="${dep.from!=null}">
                          <a href="${rootURL}/${dep.from.url}" class="model-link inside">
                            <l:icon class="${dep.from.buildStatusIconClassName} icon-sm" alt="${dep.from.iconColor.description}"/>${dep.from.displayName}</a>
                        </j:when>
                        <j:otherwise>
                          ?
                        </j:otherwise>
                      </j:choose>

                      &#x2192; <!-- right arrow -->

                      <a href="${rootURL}/${dep.to.url}" class="model-link inside">
                        <l:icon class="${dep.to.buildStatusIconClassName} icon-sm" alt="${dep.to.iconColor.description}" />${dep.to.displayName}</a>

                      (<a href="${rootURL}/${dep.project.url}changes?from=${dep.fromId}&amp;to=${dep.toId}">${%detail}</a>)
                    </li>
                  </j:forEach>
                </ol>
              </j:if>
```

\-   Would this be needed in `changesets.jelly`?

Comment (jglick) [reply core/src/main/resources/hudson/model/AbstractBuild/summary.jelly]: I am not very familiar with this but looks like something never implemented in Pipeline.

In general, for a view refactoring like this it is necessary to examine every line in all variants, figure out what it is doing and where it came from (was this a copy of an old version of something that was since updated at the source?), and decide which parts can be made common without loss of functionality and which intrinsically need customization.

File: core/src/main/resources/hudson/model/Run/index.jelly
Comment (jglick) [inline core/src/main/resources/hudson/model/Run/index.jelly:46]: This does not belong here since it references https://github.com/jenkinsci/jenkins/blob/a855a50c3196875d96858a65448db06e7ef9205f/core/src/main/java/hudson/model/AbstractBuild.java#L219-L226 for which there is no equivalent in `WorkflowRun` (or `ExternalRun`).

**Review by jglick — COMMENTED** [review]
Looks to be on the right track but there is at least one mistake here. There may be others but I have not gone through each line and tried to verify each one individually; unfortunately without [JENKINS-31203](https://issues.jenkins.io/browse/JENKINS-31203) this is a manual process and is something the author should do. (It is nice to leave self-reviews with notes for reviewers.) Should be kept in draft status until the full set of changes including in https://github.com/jenkinsci/workflow-job-plugin/pull/364 is (self-)reviewed and some sanity tests have been run on freestyle builds, Pipeline builds, external builds, and to be on the safe side maybe matrix builds (top-level & configuration) checking that UI elements mentioned in the diff are preserved.

**Comment by timja** [comment]
Have you checked all the changes now @janfaracik ?

**Comment by janfaracik** [comment]
> Have you checked all the changes now @janfaracik ?

These are the checks I've done. Had to make a couple of tweaks regarding AbstractBuild to maintain compatibility but all seems alright. 

image

**Review by timja — APPROVED** [review]
nice work great change

**Comment by janfaracik** [comment]
Any thoughts on moving this forward?

**Comment by timja** [comment]
@jenkinsci/core-pr-reviewers would someone mind taking a look please?

**Review by NotMyFault — APPROVED** [review]

**Comment by timja** [comment]
/label ready-for-merge

----

This PR is now ready for merge, after ~24 hours, we will merge it if there's no negative feedback.

Thanks!

**Comment by basil** [comment]
I have added an ATH workaround in https://github.com/jenkinsci/acceptance-test-harness/issues/1545. https://github.com/jenkinsci/acceptance-test-harness/issues/1548 tracks the removal of this workaround as part of fixing [JENKINS-73168](https://issues.jenkins.io/browse/JENKINS-73168).

**Pull request merged**
